src/endpoint.cpp

```diff
@@ -468,38 +602,77 @@ namespace oxen::quic
         log::debug(log_cat, "Connection (RID:{}) completed initial CID association", conn.reference_id());
     }
 
-    void Endpoint::associate_cid(const ngtcp2_cid* cid, Connection& conn)
+    void Endpoint::activate_cid(const ngtcp2_cid* cid, const uint8_t* token, Connection& conn)
     {
-        auto dir_str = conn.is_inbound() ? "SERVER"s : "CLIENT"s;
-        log::trace(
+        assert(in_event_loop());
+        auto qcid = quic_cid{*cid};
+        log::debug(
                 log_cat,
-                "{} associating CID:{} to {}",
-                dir_str,
-                oxenc::to_hex(cid->data, cid->data + cid->datalen),
+                "{} activating new CID:{} with reset token for {}",
+                conn.is_inbound() ? "SERVER" : "CLIENT",
+                qcid,
                 conn.reference_id());
+        associate_cid(qcid, conn);
+
+        // We only hold one reset token per connection at a time!
+        auto [it, _] = reset_token_map.emplace(gtls_reset_token::make_copy(token), qcid);
+        reset_token_lookup[qcid] = it->first;
+    }
 
+    void Endpoint::deactivate_cid(const ngtcp2_cid* cid, Connection& conn)
+    {
         assert(in_event_loop());
-        auto ccid = quic_cid{*cid};
-        conn_lookup.emplace(ccid, conn.reference_id());
-        conn.store_associated_cid(ccid);
+        auto qcid = quic_cid{*cid};
+        log::debug(
+                log_cat,
+                "{} deactivating CID:{} with reset token for {}",
+                conn.is_inbound() ? "SERVER" : "CLIENT",
+                qcid,
+                conn.reference_id());
+        dissociate_cid(qcid, conn);
+
+        if (auto it = reset_token_lookup.find(qcid); it != reset_token_lookup.end())
+        {
+            reset_token_map.erase(it->second);
+            reset_token_lookup.erase(it);
+            log::debug(log_cat, "Successfully deleted token for deactivated CID:{}", qcid);
+        }
+        else
+            log::debug(log_cat, "Could not find token corresponding to deactivated CID:{}", qcid);
     }
 
-    void Endpoint::dissociate_cid(const ngtcp2_cid* cid, Connection& conn)
+    void Endpoint::associate_cid(quic_cid qcid, Connection& conn)
     {
-        if (not cid->datalen)
-            return;
+        assert(in_event_loop());
+        log::trace(
+                log_cat, "{} associating CID:{} to {}", conn.is_inbound() ? "SERVER" : "CLIENT", qcid, conn.reference_id());
+
+        conn_lookup.emplace(qcid, conn.reference_id());
```

Comment:
Is there any possibility of a duplicate here that should be worried about?

Reply:
It should be many quic_cid's to one ConnectionID (default is 8:1), so I would guess not? Apologies if I'm misunderstanding

Reply:
Presumably there's an astronomically unlikely chance that there could be a collision, but I *think* CIDs are sent encrypted in both directions so a MITM shouldn't be able to *force* a collision, and either side of a connection intentionally making a new connection and re-using a CID is just messing up their own connection, so it should be fine.  At least, if I understand the question correctly.